sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/ProactiveOpenConnectionsProcessor.java

```diff
@@ -116,23 +127,73 @@ public void close() throws IOException {
         }
     }
 
-    public void recordOpenConnectionsAndInitCachesCompleted() {
-        this.isConcurrencySwitchFlow.set(false);
-        this.aggressivenessHint.set(ConnectionOpenFlowAggressivenessHint.DEFENSIVE);
-        this.reInstantiateOpenConnectionsPublisherAndSubscribe(true);
+    public void recordOpenConnectionsAndInitCachesCompleted(List<CosmosContainerIdentity> containerIdentities) {
+
+        endpointsUnderMonitorMapWriteLock.lock();
+        try {
+            for (CosmosContainerIdentity containerIdentity : containerIdentities) {
+                this.containersUnderOpenConnectionAndInitCaches.remove(
+                    ImplementationBridgeHelpers
+                        .CosmosContainerIdentityHelper
+                        .getCosmosContainerIdentityAccessor()
+                        .getContainerLink(containerIdentity)
+                );
+            }
+
+            // only switch to defensive mode if there is no container under openConnectionAndInitCachesFlow
+            if (this.containersUnderOpenConnectionAndInitCaches.isEmpty()) {
+                this.aggressivenessHint.set(ConnectionOpenFlowAggressivenessHint.DEFENSIVE);
+                this.reInstantiateOpenConnectionsPublisherAndSubscribe(true);
+            } else {
+                logger.debug(
+                    "Cannot switch to defensive mode as some of the containers are still under openConnectionAndInitCaches flow: [{}]",
+                    this.containersUnderOpenConnectionAndInitCaches);
+            }
+        } finally {
+            endpointsUnderMonitorMapWriteLock.unlock();
+        }
+
     }
 
-    public void recordOpenConnectionsAndInitCachesStarted() {
-        this.isConcurrencySwitchFlow.set(true);
-        this.aggressivenessHint.set(ConnectionOpenFlowAggressivenessHint.AGGRESSIVE);
-        this.reInstantiateOpenConnectionsPublisherAndSubscribe(false);
+    public void recordOpenConnectionsAndInitCachesStarted(List<CosmosContainerIdentity> cosmosContainerIdentities) {
+
+        endpointsUnderMonitorMapWriteLock.lock();
```

Comment:
Why ReadWriteLock?

All you need is synchronized (this) {
xxx
}

 if (shouldReInstantiatePublisher) {
            this.aggressivenessHint.set(ConnectionOpenFlowAggressivenessHint.AGGRESSIVE);
            this.reInstantiateOpenConnectionsPublisherAndSubscribe(false);
        }

Reply:
updated